### PR TITLE
KUBESAW-172: Restart host-operator at the end of the register-member command

### PR DESCRIPTION
**Pull request by fbm3307**
This PR is to use the `Restart` command from `Register-Member ` command and restart host operator 

There Related PR in Toolchain-E2e

Toolchain-E2E :  https://github.com/codeready-toolchain/toolchain-e2e/pull/1080

### CODE REVIEW

File: pkg/cmd/adm/register_member.go [hunk 1]
```diff
@@ -509,6 +518,20 @@ until the SpaceProvisionerConfig.spec.enabled is set to true.
 `, v.hostClusterData.apiEndpoint))
 }
 
+func (v *registerMemberValidated) getRegMemConfigFlagsAndClient(ctx *clicontext.CommandContext, clusterName string) (kubeConfigFlag *genericclioptions.ConfigFlags, rccl runtimeclient.Client, err error) {
```
Comment (mfrancisc) [inline pkg/cmd/adm/register_member.go:521]: minor - do you think it makes sense to cover this with a test?

**Review by mfrancisc — COMMENTED** [review]
Nice Job 👍 

I have only few minor comments.

File: pkg/cmd/adm/register_member.go [hunk 2]
```diff
@@ -351,6 +353,7 @@ type registerMemberValidated struct {
 	memberClusterData clusterData
 	warnings          []string
 	errors            []string
+	restart           func(ctx *clicontext.CommandContext, clusterName string, cfcGetter ConfigFlagsAndClientGetterFunc) error
```
[the remark this answers is not in the document]
Comment (fbm3307) [reply pkg/cmd/adm/register_member.go]: Updated it

File: pkg/cmd/adm/register_member.go (hunk 1 above)
Comment (fbm3307) [reply pkg/cmd/adm/register_member.go:521]: I thought about it, but couldn't find a easier way to test it, and also the similar functionality we did not test it in restart.go.

File: pkg/cmd/adm/unregister_member_test.go [hunk 3]
```diff
@@ -21,7 +21,7 @@ func TestUnregisterMemberWhenAnswerIsY(t *testing.T) {
 	ctx := clicontext.NewCommandContext(term, newClient)
 
 	// when
-	err := UnregisterMemberCluster(ctx, "member1", func(ctx *clicontext.CommandContext, clusterName string) error {
+	err := UnregisterMemberCluster(ctx, "member1", func(ctx *clicontext.CommandContext, clusterName string, cfcGetter ConfigFlagsAndClientGetterFunc) error {
```
Comment (mfrancisc) [inline pkg/cmd/adm/unregister_member_test.go:24]: minor - but I would set those params to `_` if not used
```suggestion
	err := UnregisterMemberCluster(ctx, "member1", func(_ *clicontext.CommandContext, _ string, _ ConfigFlagsAndClientGetterFunc) error {
```

Comment (mfrancisc) [reply pkg/cmd/adm/unregister_member_test.go:24]: and in the other places as well

File: pkg/cmd/adm/unregister_member_test.go [hunk 4]
```diff
@@ -65,9 +65,9 @@ func TestUnregisterMemberCallsRestart(t *testing.T) {
 	ctxAct := clicontext.NewCommandContext(term, newClient)
 	called := 0
 	// when
-	err := UnregisterMemberCluster(ctxAct, "member1", func(ctx *clicontext.CommandContext, restartClusterName string) error {
+	err := UnregisterMemberCluster(ctxAct, "member1", func(ctx *clicontext.CommandContext, restartClusterName string, cfcGetter ConfigFlagsAndClientGetterFunc) error {
```
Comment (mfrancisc) [inline pkg/cmd/adm/unregister_member_test.go:68]: same here , the cfGetter is not used
```suggestion
	err := UnregisterMemberCluster(ctxAct, "member1", func(ctx *clicontext.CommandContext, restartClusterName string, _ ConfigFlagsAndClientGetterFunc) error {
```

File: pkg/cmd/adm/unregister_member_test.go [hunk 5]
```diff
@@ -65,9 +65,9 @@ func TestUnregisterMemberCallsRestart(t *testing.T) {
...
 		called++
-		return mockRestart(ctx, restartClusterName)
+		return mockRestart(ctx, restartClusterName, getConfigFlagsAndClient)
```
Comment (mfrancisc) [inline pkg/cmd/adm/unregister_member_test.go:70]: it is not actually called so we can just pass nil right ? 
```suggestion
		return mockRestart(ctx, restartClusterName, nil)
```

File: pkg/cmd/adm/unregister_member_test.go [hunk 6]
```diff
@@ -164,9 +164,13 @@ func TestUnregisterMemberLacksPermissions(t *testing.T) {
...
 
-func mockRestart(ctx *clicontext.CommandContext, clusterName string) error {
+func mockRestart(ctx *clicontext.CommandContext, clusterName string, cfc ConfigFlagsAndClientGetterFunc) error {
 	if clusterName == "host" && ctx != nil {
 		return nil
 	}
+	_, _, err := cfc(ctx, clusterName)
+	if err == nil {
+		return nil
+	}
```
Comment (mfrancisc) [inline pkg/cmd/adm/unregister_member_test.go:174]: this part is never being called if I'm not missing something

**Review by mfrancisc — COMMENTED** [review]
Just few minor things related to the tests

File: pkg/cmd/adm/register_member.go [hunk 7]
```diff
@@ -483,6 +486,11 @@ func (v *registerMemberValidated) perform(ctx *extendedCommandContext) error {
 		return err
 	}
 
+	// restart Host Operator using the adm-restart command
+	if err := v.restartFunc(ctx.CommandContext, "host", v.getRegMemConfigFlagsAndClient); err != nil {
```
[the remark this answers is not in the document]
Comment (fbm3307) [reply pkg/cmd/adm/register_member.go]: updated it to `configuration.hostname` see [a05065e](https://github.com/kubesaw/ksctl/pull/90/commits/a05065ec42b65641228054641e05b47a2fe10450)

File: pkg/cmd/adm/unregister_member_test.go (hunk 3 above)
Comment (fbm3307) [reply pkg/cmd/adm/unregister_member_test.go]: updated it to all places , see [a05065e](https://github.com/kubesaw/ksctl/pull/90/commits/a05065ec42b65641228054641e05b47a2fe10450)

File: pkg/cmd/adm/unregister_member_test.go (hunk 4 above)
Comment (fbm3307) [reply pkg/cmd/adm/unregister_member_test.go]: yes right, updated it [a05065e](https://github.com/kubesaw/ksctl/pull/90/commits/a05065ec42b65641228054641e05b47a2fe10450)

File: pkg/cmd/adm/unregister_member_test.go (hunk 5 above)
Comment (fbm3307) [reply pkg/cmd/adm/unregister_member_test.go:70]: no i want to check the `getConfigFlagsAndClient` func  in mock restart, that it just give error

File: pkg/cmd/adm/unregister_member_test.go (hunk 6 above)
Comment (fbm3307) [reply pkg/cmd/adm/unregister_member_test.go]: updated the logic

**Review by mfrancisc — APPROVED** [review]
Nice Job 👍 

Thanks for addressing my comments.

File: pkg/cmd/adm/unregister_member_test.go [hunk 8]
```diff
@@ -164,9 +164,11 @@ func TestUnregisterMemberLacksPermissions(t *testing.T) {
 	AssertToolchainClusterSpec(t, fakeClient, toolchainCluster)
 }
 
-func mockRestart(ctx *clicontext.CommandContext, clusterName string) error {
-	if clusterName == "host" && ctx != nil {
+func mockRestart(ctx *clicontext.CommandContext, clusterName string, cfc ConfigFlagsAndClientGetterFunc) error {
+	_, _, cfcerr := cfc(ctx, clusterName)
+	if clusterName == "host" && ctx != nil && cfcerr == nil {
```
Comment (mfrancisc) [inline pkg/cmd/adm/unregister_member_test.go:169]: do we event need to check for `clusterName == "host" ` , it's only being called with `member1` as an argument If I'm seeing it correctly.

Comment (fbm3307) [reply pkg/cmd/adm/unregister_member_test.go:169]: its just checking that when restart is being called, cluster name which is being passed is host

File: pkg/cmd/adm/register_member.go [hunk 9]
```diff
@@ -509,6 +517,20 @@ until the SpaceProvisionerConfig.spec.enabled is set to true.
 `, v.hostClusterData.apiEndpoint))
 }
 
+func (v *registerMemberValidated) getRegMemConfigFlagsAndClient(ctx *clicontext.CommandContext, clusterName string) (kubeConfigFlag *genericclioptions.ConfigFlags, rccl runtimeclient.Client, err error) {
```
Comment (MatousJobanek) [inline pkg/cmd/adm/register_member.go:520]: the params are not used: 
```suggestion
func (v *registerMemberValidated) getRegMemConfigFlagsAndClient(_ *clicontext.CommandContext, _ string) (kubeConfigFlag *genericclioptions.ConfigFlags, rccl runtimeclient.Client, err error) {
```

File: pkg/cmd/adm/register_member.go [hunk 10]
```diff
@@ -94,8 +95,8 @@ func NewRegisterMemberCmd() *cobra.Command {
 	return cmd
 }
 
-func registerMemberCluster(ctx *extendedCommandContext, args registerMemberArgs) error {
-	validated, err := validateArgs(ctx, args)
+func registerMemberCluster(ctx *extendedCommandContext, args registerMemberArgs, restart restartFunc) error {
+	validated, err := validateArgs(ctx, args, restart)
```
[the remark this answers is not in the document]
Comment (fbm3307) [reply pkg/cmd/adm/register_member.go]: Okay thanks for explaining me during the call, i have updated it now .

File: pkg/cmd/adm/register_member_test.go [hunk 11]
```diff
@@ -82,7 +83,9 @@ func TestRegisterMember(t *testing.T) {
 		}
 
 		// when
-		err := registerMemberCluster(ctx, newRegisterMemberArgsWith(hostKubeconfig, memberKubeconfig, false))
+		err := registerMemberCluster(ctx, newRegisterMemberArgsWith(hostKubeconfig, memberKubeconfig, false), func(_ *clicontext.CommandContext, _ string, _ ConfigFlagsAndClientGetterFunc) error {
+			return nil
```
[the remark this answers is not in the document]
Comment (fbm3307) [reply pkg/cmd/adm/register_member_test.go:87]: added it

**Review by MatousJobanek — APPROVED** [review]
Looks good, thanks

**Pull request merged**